src/python/gudhi/barycenter.py

```diff
@@ -0,0 +1,227 @@
+import ot
+import numpy as np
+import scipy.spatial.distance as sc
+
+
+# This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
+# See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
+# Author(s):       Theo Lacombe
+#
+# Copyright (C) 2019 Inria
+#
+# Modification(s):
+#   - YYYY/MM Author: Description of the modification
+
+
+def _proj_on_diag(w):
+    '''
+        Util function to project a point on the diag.
+    '''
+    return np.array([(w[0] + w[1])/2 , (w[0] + w[1])/2])
+
+
+def _proj_on_diag_array(X):
+    '''
+    :param X: (n x 2) array encoding the points of a persistent diagram.
+    :returns: (n x 2) array encoding the (respective orthogonal) projections of the points onto the diagonal
+    '''
+    Z = (X[:,0] + X[:,1]) / 2.
+    return np.array([Z , Z]).T
+
+
+def _build_dist_matrix(X, Y, p=2., q=2.):
+    '''
+    :param X: (n x 2) numpy.array encoding the (points of the) first diagram.
+    :param Y: (m x 2) numpy.array encoding the second diagram.
+    :param q: Ground metric (i.e. norm l_q).
+    :param p: exponent for the Wasserstein metric.
+    :returns: (n+1) x (m+1) np.array encoding the cost matrix C. 
+                For 1 <= i <= n, 1 <= j <= m, C[i,j] encodes the distance between X[i] and Y[j], while C[i, m+1] (resp. C[n+1, j]) encodes the distance (to the p) between X[i] (resp Y[j]) and its orthogonal proj onto the diagonal.
+                note also that C[n+1, m+1] = 0  (it costs nothing to move from the diagonal to the diagonal).
+                Note that for lagrangian_barycenter, one must use p=q=2. 
+    '''
+    Xdiag = _proj_on_diag_array(X)
+    Ydiag = _proj_on_diag_array(Y)
+    if np.isinf(q):
+        C = sc.cdist(X, Y, metric='chebyshev')**p
+        Cxd = np.linalg.norm(X - Xdiag, ord=q, axis=1)**p
+        Cdy = np.linalg.norm(Y - Ydiag, ord=q, axis=1)**p
```

Comment:
Actually, the current algorithm only works if p=q=2 (btw, I should rename these variables with respect to the name we recently chose for `wasserstein_distance.py` ...!)

Allowing for different `p,q` is just a preliminary step for further algorithms (working differently, and should work with other values for `p,q`). This shouldn't be used for now. In the current implementation, the user shouldn't be able to change it as these are not parameters of `lagrangian_barycenter` (they are only used in intermediate functions, like `_optimal_matching`, with default parameters).
Would it be safer to remove it?

Reply:
If it is some dead code, it is safer to remove it, thanks. If not, please test it.
Ok for the renaming.

Reply:
Well I wouldn't say it is some dead code. It is rather some "not-born-yet" code ; in sense that I will need it in upcoming implementations of barycenter estimation with different approaches. But it is not used yet.

But ok, I will rewrite it for now (restricting to the case p=q=2), and will reintroduce this version when implementing other algorithms.

Reply:
From Vincent's comment, it seems that he would also be happy if you kept the code, and just added a very small test that calling _build_dist_matrix returns what is expected (the notion of unit test means testing some private functions as well, not just the final public one). Also, just moving this code to wasserstein.py (or importing stuff from wasserstein.py) would mean that this isn't dead code anymore, since the duplicate function there is used, no?

Reply:
Yes I could do that and that would be the proper way to go. When I started implementing this, `wasserstein.py` wasn't merge to `master`. I think I actually need a bit of re-writing for this code overall.

Reply:
I finally removed _build_dist_matrix from this file and imported the one of `wasserstein.py`.